File: mobile_client/lib/services/game/game_manager_service.dart
import 'package:flutter/material.dart';
import 'package:mobile_client/constants/enum/page_url.dart';
import 'package:mobile_client/constants/enum/room_destruction_reason.dart';
import 'package:mobile_client/constants/socket_events/game_event.dart';
import 'package:mobile_client/constants/socket_events/room_event.dart';
import 'package:mobile_client/globals.dart';
import 'package:mobile_client/models/player_info.dart';
import 'package:mobile_client/models/question.dart';
import 'package:mobile_client/services/game/game_state_service.dart';
import 'package:mobile_client/services/game/prize_service.dart';
import 'package:mobile_client/services/game/review_manager_service.dart';
import 'package:mobile_client/services/game/submit_manager_service.dart';
import 'package:mobile_client/services/websocket_service.dart';
import 'package:mobile_client/utils/app_navigator.dart';
import 'package:mobile_client/widgets/snackbar.dart';

class GameManagerService {
  final GameStateService _gameStateService;
  final WebSocketService _webSocketService;
  final SubmitManagerService _submitManagerService;
  final ReviewManagerService _reviewManager;
  final PrizeService _prizeService;

  GameManagerService(
    this._gameStateService,
    this._webSocketService,
    this._submitManagerService,
    this._reviewManager,
    this._prizeService,
  );

  Future<void> submitAnswer() async {
    if (_gameStateService.isSubmitted.value ||
        _gameStateService.isEliminated.value) {
      return;
    }
    _gameStateService.submit();
    final playerAnswer =
        await _submitManagerService.getAnswer(_gameStateService.questionType);
    _webSocketService.send(GameEvent.submitAnswer.name, playerAnswer);
  }

  void setupManager() {
    _setupDisconnectClientsEvent();
    _setupMoveToNextQuestionEvent();
    _setGameFinishedEvent();
    _setupEliminatedPlayer();
    _reviewManager.setupManager();
    _prizeService.initialize();
  }

  void resetManager() {
    _gameStateService.resetGameState();
    _reviewManager.resetReviewManager();
    _prizeService.removeListeners();
    _removeManagerListeners();

    if (_webSocketService.isSocketConnected()) {
      _webSocketService.send(RoomEvent.leaveGame.name);
    }
  }

  void _removeManagerListeners() {
    _webSocketService.removeAllListeners(GameEvent.nextQuestion.name);
    _webSocketService.removeAllListeners(GameEvent.playerEliminated.name);
    _webSocketService.removeAllListeners(GameEvent.gameFinished.name);
    _webSocketService.removeAllListeners(RoomEvent.leaveGame.name);
  }

  void _setupMoveToNextQuestionEvent() {
    _webSocketService.on<Map<String, dynamic>?>(GameEvent.nextQuestion.name,
        (json) {
      if (json == null) return;
      final question = Question.fromJson(json);
      _gameStateService.changeQuestion(question);
      _submitManagerService.setInputs(question);
    });
  }

  void _setGameFinishedEvent() {
    _webSocketService.once<List<dynamic>>(GameEvent.gameFinished.name, (data) {
      final playerInfos =
          data?.map((item) => PlayerInfo.fromJson(item)).toList();
      _gameStateService.playerInfos = playerInfos ?? [];
      AppNavigator.go('/${PageUrl.results.name}');
    });
  }

  void _setupEliminatedPlayer() {
    _webSocketService.once(GameEvent.playerEliminated.name, (_) {
      _gameStateService.isEliminated.value = true;
      if (snackbarKey.currentContext is BuildContext &&
          snackbarKey.currentContext!.mounted) {
        showCustomSnackBar(
            context: snackbarKey.currentContext!,
            message: 'Vous avez été éliminé de la partie');
      }
    });
  }

  void _setupDisconnectClientsEvent() {
    _webSocketService.once<dynamic>(RoomEvent.leaveGame.name, (data) {
      final roomDestructionReason = RoomDestructionReason.fromJson(data);
      switch (roomDestructionReason) {
        case RoomDestructionReason.allPlayerLeft:
          showCustomGlobalSnackBar(
              message: 'Tous les joueurs ont quitté la partie');
          break;
        case RoomDestructionReason.organizerLeft:
          showCustomGlobalSnackBar(
              message: "L'organisateur a quitté ou a mis fin à la partie");
        default:
          break;
      }
      AppNavigator.go('/${PageUrl.home.name}');
    });
  }
}

File: desktop-client/src/app/game/services/game-manager/game-manager.service.ts
import { Injectable } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Router } from '@angular/router';
import { PageUrl } from '@app/enum/page-url';
import { GameStateService } from '@app/game/services/game-state/game-state.service';
import { PrizeService } from '@app/game/services/prize/prize.service';
import { ReviewManagerService } from '@app/game/services/review-manager/review-manager.service';
import { SubmitManagerService } from '@app/game/services/submit-manager/submit-manager.service';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { RoomDestructionReason } from '@common/enum/room-destruction-reason';
import { PlayerInfo } from '@common/interfaces/player-info';
import { Question } from '@common/interfaces/question';
import { GameEvent } from '@common/socket-events/game-event';
import { RoomEvent } from '@common/socket-events/room-event';
import { BehaviorSubject, Observable } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class GameManagerService {
    private loadNextQuestionSubject: BehaviorSubject<void> = new BehaviorSubject<void>(undefined);

    // eslint-disable-next-line max-params
    constructor(
        private readonly gameStateService: GameStateService,
        private readonly webSocketService: WebsocketService,
        private readonly router: Router,
        private readonly submitManagerService: SubmitManagerService,
        private readonly reviewManager: ReviewManagerService,
        private readonly prizeService: PrizeService,
        private readonly snackBar: MatSnackBar,
    ) {}

    get loadNextQuestionObservable(): Observable<void> {
        return this.loadNextQuestionSubject.asObservable();
    }

    async submitAnswer() {
        if (this.gameStateService.isSubmitted || this.gameStateService.isEliminated) return;
        this.gameStateService.isSubmitted = true;
        const playerAnswer = await this.submitManagerService.getAnswer(this.gameStateService.questionType);
        this.webSocketService.send(GameEvent.SUBMIT_ANSWER, playerAnswer);
    }

    setupManager() {
        this.setupDisconnectClientsEvent();
        this.setupMoveToNextQuestionEvent();
        this.setGameFinishedEvent();
        this.setupEliminatedPlayer();
        this.reviewManager.setupManager();
        this.prizeService.initialize();
    }

    resetManager() {
        this.gameStateService.resetGameState();
        this.reviewManager.resetReviewManager();
        this.removeManagerListeners();
        this.prizeService.removeListeners();
        if (this.webSocketService.isSocketConnected()) this.webSocketService.send(RoomEvent.LEAVE_GAME);
    }

    private removeManagerListeners() {
        this.webSocketService.removeAllListeners(GameEvent.NEXT_QUESTION);
        this.webSocketService.removeAllListeners(GameEvent.GAME_FINISHED);
        this.webSocketService.removeAllListeners(RoomEvent.LEAVE_GAME);
        this.webSocketService.removeAllListeners(GameEvent.PLAYER_ELIMINATED);
    }

    private setupMoveToNextQuestionEvent() {
        this.webSocketService.on(GameEvent.NEXT_QUESTION, (question?: Question) => {
            if (!question) return;
            this.gameStateService.changeQuestion(question);
            this.submitManagerService.setInputs(question);
            this.loadNextQuestionSubject.next();
        });
    }

    private setGameFinishedEvent() {
        this.webSocketService.once<PlayerInfo[]>(GameEvent.GAME_FINISHED, (playerInfo?: PlayerInfo[]) => {
            this.gameStateService.playerInfo = playerInfo ?? [];
            this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.RESULTS}`]);
        });
    }

    private setupDisconnectClientsEvent() {
        this.webSocketService.once(RoomEvent.LEAVE_GAME, (roomDestructionReason?: RoomDestructionReason) => {
            switch (roomDestructionReason) {
                case RoomDestructionReason.ALL_PLAYER_LEFT:
                    this.snackBar.open('Tous les joueurs ont quitté la partie');
                    break;
                case RoomDestructionReason.ORGANIZER_LEFT:
                    this.snackBar.open("L'organisateur a quitté ou a mis fin à la partie");
                    break;
            }
            this.router.navigate([`/${PageUrl.APP_PREFIX}/${PageUrl.HOME}`]);
        });
    }

    private setupEliminatedPlayer() {
        this.webSocketService.once(GameEvent.PLAYER_ELIMINATED, () => {
            this.gameStateService.isEliminated = true;
            this.snackBar.open('Vous avez été éliminé de la partie');
        });
    }
}
